File: src/com/app/test/redenveloped/RedEnvelopesHelper.java
package com.app.test.redenveloped;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.PropertyValuesHolder;
import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Matrix;
import android.support.annotation.IntDef;
import android.view.animation.LinearInterpolator;

import com.app.test.R;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

public class RedEnvelopesHelper {

    public interface RedPacketInter {
        void clickRedPacket(RedEnvelopesHelper bean);

        Bitmap getNormalBitmap();

        Bitmap getClickBitmap();

        Bitmap getGoldBitmap();
    }

    private RedPacketInter redPacketInter;

    public RedPacketInter getRedPacketInter() {
        return redPacketInter;
    }

    public void setRedPacketInter(RedPacketInter redPacketInter) {
        this.redPacketInter = redPacketInter;
    }

    private RedEnvelopes redPacketBean;
    private Context context;
    private Integer index = -1;
    private int money;
    //x,y代表当前红包图片绘制位置
    private int x;
    private int y;
    //初始化时的Y坐标
    private int initY;
    //点击时的x,y坐标，防止动画执行刷新x，y坐标值，在做逻辑判断时提前记录
    private int currentX;
    private int currentY;
    private boolean canClick = true;
    private int normalWidth;
    private int normalHeight;
    private int clickBitmapWidth;
    private int clickBitmapHeight;

    private int goldBitmapWidth;
    private int goldBitmapHeight;

    private long durationTime = 3000;
    private long showDismissTime = 500;
    //金币移动动画执行时间
    private int moneyMoveTime = 800;

    private ValueAnimator valueAnimator;
    private boolean isEnd;
    private Matrix matrix;
    //3：红包下落,2:显示被点击红包，1显示金币，0：消失
    private int redPacketStatus = 3;


    public static final int status_3 = 3;
    public static final int status_2 = 2;
    public static final int status_1 = 1;
    public static final int status_0 = 0;

    @IntDef({status_3, status_2, status_1, status_0})
    @Retention(RetentionPolicy.SOURCE)
    public @interface status {
    }

    //金币消失x方向偏移量
    private int goldDismissXOffset;
    //金币消失y方向偏移量
    private int goldDismissYOffset;

    public void setGoldDismissXOffset(int goldDismissXOffset) {
        this.goldDismissXOffset = goldDismissXOffset;
    }

    public void setGoldDismissYOffset(int goldDismissYOffset) {
        this.goldDismissYOffset = goldDismissYOffset;
    }

    public int getGoldDismissXOffset() {
        return goldDismissXOffset;
    }

    public int getGoldDismissYOffset() {
        return goldDismissYOffset;
    }

    public Integer getIndex() {
        return index;
    }

    public void setIndex(Integer index) {
        this.index = index;
    }

    public int getMoney() {
        return money;
    }

    public void setMoney(int money) {
        this.money = money;
    }

    public int getRedPacketStatus() {
        return redPacketStatus;
    }

    public void setRedPacketStatus(@status int redPacketStatus) {
        this.redPacketStatus = redPacketStatus;
    }

    public RedEnvelopesHelper(Context context) {
        this.context = context;
    }

    public boolean isEnd() {
        return isEnd;
    }

    public void setEnd(boolean end) {
        isEnd = end;
    }

    public int getX() {
        return x;
    }

    public void setX(int x) {
        this.x = x;
    }

    public int getY() {
        return y;
    }

    public void setY(int y) {
        this.y = y;
    }

    public void setInitY(int initY) {
        this.initY = initY;
        setY(initY);
    }

    public void saveCurrentXY() {
        currentX = getX();
        currentY = getY();
    }

    public Bitmap getBitmap() {
        return getRedPacketInter() == null ? null : getRedPacketInter().getNormalBitmap();
    }

    public Bitmap getClickBitmap() {
        return getRedPacketInter() == null ? null : getRedPacketInter().getClickBitmap();
    }

    public Bitmap getGoldBitmap() {
        return getRedPacketInter() == null ? null : getRedPacketInter().getGoldBitmap();
    }

    public int getNormalWidth() {
        return normalWidth;
    }

    public void setNormalWidth(int width) {
        this.normalWidth = width;
    }

    public int getNormalHeight() {
        return normalHeight;
    }

    public void setNormalHeight(int height) {
        this.normalHeight = height;
    }

    public int getClickBitmapWidth() {
        return clickBitmapWidth;
    }

    public void setClickBitmapWidth(int clickBitmapWidth) {
        this.clickBitmapWidth = clickBitmapWidth;
    }

    public int getClickBitmapHeight() {
        return clickBitmapHeight;
    }

    public void setClickBitmapHeight(int clickBitmapHeight) {
        this.clickBitmapHeight = clickBitmapHeight;
    }

    public int getGoldBitmapWidth() {
        return goldBitmapWidth;
    }

    public void setGoldBitmapWidth(int goldBitmapWidth) {
        this.goldBitmapWidth = goldBitmapWidth;
    }

    public int getGoldBitmapHeight() {
        return goldBitmapHeight;
    }

    public void setGoldBitmapHeight(int goldBitmapHeight) {
        this.goldBitmapHeight = goldBitmapHeight;
    }

    public Matrix getMatrix() {
        return matrix == null ? new Matrix() : matrix;
    }

    public boolean isCanClick() {
        return canClick;
    }

    public boolean constant(int x, int y) {
        if (x >= currentX && x <= currentX + getNormalWidth() && y >= currentY && y <= getNormalHeight() + currentY) {
            return true;
        }
        return false;
    }

    public void setDurationTime(long time) {
        if (time <= 0) {
            time = 3000;
        }
        durationTime = time;
    }


    public void startDown(int containerHeight) {
        ValueAnimator valueAnimator = ValueAnimator.ofFloat(initY, containerHeight);
        valueAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                if (getRedPacketStatus() != status_3) {
                    return;
                }
                float animatedValue = (Float) animation.getAnimatedValue();
                setY((int) animatedValue);
            }
        });
        valueAnimator.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationStart(Animator animation) {
                setRedPacketStatus(status_3);
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                super.onAnimationEnd(animation);
                if (canClick) {
                    setEnd(true);
                }
            }
        });
        valueAnimator.setDuration(durationTime);
        valueAnimator.setRepeatCount(0);
        valueAnimator.setInterpolator(new LinearInterpolator());
        valueAnimator.start();
    }

    private void startChangeToRedPacket() {
        //防止重复点击
        canClick = false;
        if (valueAnimator != null && valueAnimator.isRunning()) {
            valueAnimator.cancel();
        }
        final int scaleX;
        final int scaleY;
        final int translateX;

        scaleX = getClickBitmapWidth() / 2;
        scaleY = getClickBitmapHeight() / 2;
        translateX = scaleX - getNormalWidth() / 2;

        valueAnimator = ValueAnimator.ofFloat(0.8f, 1.2f);
        valueAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                if (getRedPacketStatus() != status_2) {
                    return;
                }
                if (matrix == null) {
                    matrix = new Matrix();
                } else {
                    matrix.reset();
                }
                float animatedValue = (float) animation.getAnimatedValue();
                matrix.setTranslate(getX() - translateX, getY() - 20);
                matrix.preScale(animatedValue, animatedValue, scaleX, scaleY);
            }
        });
        valueAnimator.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationStart(Animator animation) {
                setRedPacketStatus(status_2);
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                setRedPacketStatus(status_1);
                startChangeToGold();
            }
        });
        valueAnimator.setDuration(showDismissTime);
        valueAnimator.setRepeatCount(0);
        valueAnimator.setInterpolator(new LinearInterpolator());
        valueAnimator.start();
    }

    private void startChangeToGold() {
        if (valueAnimator != null && valueAnimator.isRunning()) {
            valueAnimator.cancel();
        }
        final Bitmap bitmap = BitmapFactory.decodeResource(context.getResources(), R.drawable.ic_redpacket_gold_active);
        final int scaleX;
        final int scaleY;
        if (bitmap == null) {
            scaleX = 150 / 2;
            scaleY = 150 / 2;
        } else {
            scaleX = bitmap.getWidth() / 2;
            scaleY = bitmap.getHeight() / 2;
        }
//        setBitmap(bitmap);
        PropertyValuesHolder holder1 = PropertyValuesHolder.ofFloat("x", getX(), getGoldDismissXOffset() - scaleX / 2);
        PropertyValuesHolder holder2 = PropertyValuesHolder.ofFloat("y", getY(), getGoldDismissYOffset());
        PropertyValuesHolder holder3 = PropertyValuesHolder.ofFloat("scale", 0.8f, 0.5f);
        valueAnimator = ValueAnimator.ofPropertyValuesHolder(holder1, holder2, holder3);
        valueAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                if (getRedPacketStatus() != status_1) {
                    return;
                }
                float x = (float) animation.getAnimatedValue("x");
                float y = (float) animation.getAnimatedValue("y");
                float scale = (float) animation.getAnimatedValue("scale");
                if (matrix == null) {
                    matrix = new Matrix();
                } else {
                    matrix.reset();
                }
                matrix.setTranslate(getX() + scaleX / 3, getY());
                matrix.preScale(scale, scale, scaleX, scaleY);
                setX((int) x);
                setY((int) y);
            }
        });
        valueAnimator.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationStart(Animator animation) {
                setRedPacketStatus(status_1);
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                setRedPacketStatus(status_0);
                setEnd(true);
            }
        });
        valueAnimator.setDuration(moneyMoveTime);
        valueAnimator.setRepeatCount(0);
        valueAnimator.setInterpolator(new LinearInterpolator());
        valueAnimator.start();
    }

    public void clickRedPacket() {
        if (getRedPacketInter() != null) {
            getRedPacketInter().clickRedPacket(this);
        }
        startChangeToRedPacket();
    }


    public static RedEnvelopesHelper produceRedPacket(int index, Context context, int normalBitmapWidth, int normalBitmapHeight, int clickBitmapWidth, int clickBitmapHeight, int redPacketX, long durationTimeMillis) {
        if (context == null) {
            return null;
        }
        RedEnvelopesHelper redPacket = new RedEnvelopesHelper(context);
//        int i = getMaxMillis()- temp* (checkCount - 1);
        redPacket.setDurationTime(durationTimeMillis);
        redPacket.setIndex(index);

        redPacket.setNormalWidth(normalBitmapWidth);
        redPacket.setNormalHeight(normalBitmapHeight);

        redPacket.setClickBitmapWidth(clickBitmapWidth);
        redPacket.setClickBitmapHeight(clickBitmapHeight);


//            redPacket.setBitmap(redPacketBitmap);


        redPacket.setX(redPacketX);
        redPacket.setInitY(-redPacket.getNormalHeight());

//        int nextRedPacketX = getNextRedPacketX(activity,screenWidth, redPacket.getNormalWidth());
//        rpv.addRedPacketData(checkCount-1,redPacket);
//        redPacket.startDown(RedEnvelopesHelper.getScreenHeight(activity));
        return redPacket;
    }


}

File: src/com/app/test/redenveloped/RedEnvelopesHelper.kt
package com.app.test.redenveloped

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.PropertyValuesHolder
import android.animation.ValueAnimator
import android.animation.ValueAnimator.AnimatorUpdateListener
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.view.animation.LinearInterpolator
import com.app.test.R

class RedEnvelopesHelper(private val context: Context) {

    interface RedPacketLister {
        fun clickRedPacket(bean: RedEnvelopesHelper?)
        fun returnNormalBitmap(): Bitmap?
        fun returnClickBitmap(): Bitmap?
        fun returnGoldBitmap(): Bitmap?
    }

    var redPacketLister: RedPacketLister? = null

    var index = -1
    var money = 0

    //x,y代表当前红包图片绘制位置
    var moveX = 0
    var moveY = 0

    //初始化时的Y坐标
    private var initY = 0

    //点击时的x,y坐标，防止动画执行刷新x，y坐标值，在做逻辑判断时提前记录
    private var currentX = 0
    private var currentY = 0
    var isCanClick = true
        private set
    var normalWidth = 0
    var normalHeight = 0
    var clickBitmapWidth = 0
    var clickBitmapHeight = 0
    private var durationTime: Long = 3000
    private val showDismissTime: Long = 500

    //金币移动动画执行时间
    private val moneyMoveTime = 800
    private var valueAnimator: ValueAnimator? = null
    var isEnd = false
    private var matrix: Matrix? = null

    //3：红包下落,2:显示被点击红包，1显示金币，0：消失
    var redPacketStatus = 3

//    @IntDef(status_3, status_2, status_1, status_0)
//    @kotlin.annotation.Retention(AnnotationRetention.SOURCE)
//    annotation class status

    //金币消失x方向偏移量
    var goldDismissXOffset = 0

    //金币消失y方向偏移量
    var goldDismissYOffset = 0

    fun setInitY(initY: Int) {
        this.initY = initY
        moveY = initY
    }

    fun saveCurrentXY() {
        currentX = moveX
        currentY = moveY
    }

    val bitmap: Bitmap?
        get() = if (redPacketLister == null) null else redPacketLister?.returnNormalBitmap()

    val clickBitmap: Bitmap?
        get() = if (redPacketLister == null) null else redPacketLister?.returnClickBitmap()

    val goldBitmap: Bitmap?
        get() = if (redPacketLister == null) null else redPacketLister?.returnGoldBitmap()

    fun getMatrix(): Matrix {
        return (if (matrix == null) Matrix() else matrix!!)
    }

    fun constant(x: Int, y: Int): Boolean {
        return x >= currentX && x <= currentX + normalWidth && y >= currentY && y <= normalHeight + currentY
    }

    fun setDurationTime(time: Long) {
        var t = time
        if (t <= 0) {
            t = 3000
        }
        durationTime = t
    }

    fun startDown(containerHeight: Int) {
        val valueAnimator = ValueAnimator.ofFloat(initY.toFloat(), containerHeight.toFloat())
        valueAnimator.addUpdateListener(AnimatorUpdateListener { animation ->
            if (redPacketStatus != status_3) {
                return@AnimatorUpdateListener
            }
            val animatedValue = animation.animatedValue as Float
            moveY = animatedValue.toInt()
        })
        valueAnimator.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationStart(animation: Animator) {
                redPacketStatus = status_3
            }

            override fun onAnimationEnd(animation: Animator) {
                super.onAnimationEnd(animation)
                if (isCanClick) {
                    isEnd = true
                }
            }
        })
        valueAnimator.duration = durationTime
        valueAnimator.repeatCount = 0
        valueAnimator.interpolator = LinearInterpolator()
        valueAnimator.start()
    }

    private fun startChangeToRedPacket() {
        //防止重复点击
        isCanClick = false
        if (valueAnimator?.isRunning == true) {
            valueAnimator?.cancel()
        }
        val translateX: Int
        val scaleX: Int = clickBitmapWidth / 2
        val scaleY: Int = clickBitmapHeight / 2
        translateX = scaleX - normalWidth / 2
        valueAnimator = ValueAnimator.ofFloat(0.8f, 1.2f)
        valueAnimator?.addUpdateListener(AnimatorUpdateListener { animation ->
            if (redPacketStatus != status_2) {
                return@AnimatorUpdateListener
            }
            if (matrix == null) {
                matrix = Matrix()
            } else {
                matrix?.reset()
            }
            val animatedValue = animation.animatedValue as Float
            matrix?.setTranslate(moveX - translateX.toFloat(), moveY - 20.toFloat())
            matrix?.preScale(animatedValue, animatedValue, scaleX.toFloat(), scaleY.toFloat())
        })
        valueAnimator?.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationStart(animation: Animator) {
                redPacketStatus = status_2
            }

            override fun onAnimationEnd(animation: Animator) {
                redPacketStatus = status_1
                startChangeToGold()
            }
        })
        valueAnimator?.duration = showDismissTime
        valueAnimator?.repeatCount = 0
        valueAnimator?.interpolator = LinearInterpolator()
        valueAnimator?.start()
    }

    private fun startChangeToGold() {
        if (valueAnimator?.isRunning == true) {
            valueAnimator?.cancel()
        }
        val bitmap = BitmapFactory.decodeResource(context.resources, R.drawable.ic_redpacket_gold_active)
        val scaleX: Int
        val scaleY: Int
        if (bitmap == null) {
            scaleX = 150 / 2
            scaleY = 150 / 2
        } else {
            scaleX = bitmap.width / 2
            scaleY = bitmap.height / 2
        }
        //        setBitmap(bitmap);
        val holder1 = PropertyValuesHolder.ofFloat("x", moveX.toFloat(), goldDismissXOffset - scaleX / 2.toFloat())
        val holder2 = PropertyValuesHolder.ofFloat("y", moveY.toFloat(), goldDismissYOffset.toFloat())
        val holder3 = PropertyValuesHolder.ofFloat("scale", 0.8f, 0.5f)
        valueAnimator = ValueAnimator.ofPropertyValuesHolder(holder1, holder2, holder3)
        valueAnimator?.addUpdateListener(AnimatorUpdateListener { animation ->
            if (redPacketStatus != status_1) {
                return@AnimatorUpdateListener
            }
            val xPosition = animation.getAnimatedValue("x") as Float
            val yPosition = animation.getAnimatedValue("y") as Float
            val scale = animation.getAnimatedValue("scale") as Float
            if (matrix == null) {
                matrix = Matrix()
            } else {
                matrix?.reset()
            }
            matrix?.setTranslate(moveX + scaleX / 3.toFloat(), moveY.toFloat())
            matrix?.preScale(scale, scale, scaleX.toFloat(), scaleY.toFloat())
            moveX = xPosition.toInt()
            moveY = yPosition.toInt()
        })
        valueAnimator?.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationStart(animation: Animator) {
                redPacketStatus = status_1
            }

            override fun onAnimationEnd(animation: Animator) {
                redPacketStatus = status_0
                isEnd = true
            }
        })
        valueAnimator?.duration = moneyMoveTime.toLong()
        valueAnimator?.repeatCount = 0
        valueAnimator?.interpolator = LinearInterpolator()
        valueAnimator?.start()
    }

    fun clickRedPacket() {
        redPacketLister?.clickRedPacket(this)
        startChangeToRedPacket()
    }

    companion object {
        const val status_3 = 3
        const val status_2 = 2
        const val status_1 = 1
        const val status_0 = 0
        fun produceRedPacket(index: Int, context: Context?, normalBitmapWidth: Int, normalBitmapHeight: Int, clickBitmapWidth: Int, clickBitmapHeight: Int, redPacketX: Int, durationTimeMillis: Long): RedEnvelopesHelper? {
            if (context == null) {
                return null
            }
            val redPacket = RedEnvelopesHelper(context)
            redPacket.setDurationTime(durationTimeMillis)
            redPacket.index = index
            redPacket.normalWidth = normalBitmapWidth
            redPacket.normalHeight = normalBitmapHeight
            redPacket.clickBitmapWidth = clickBitmapWidth
            redPacket.clickBitmapHeight = clickBitmapHeight
            redPacket.moveX = redPacketX
            redPacket.setInitY(-redPacket.normalHeight)
            return redPacket
        }
    }

}